### PR TITLE
Change the DNMD APIs to act on single rows and bring other perf improvements from the CoreCLR + DNMD experiment to DNMD.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This PR changes the various DNMD column reading/writing APIs to act primarily on 1 row instead of being designed primarily for bulk processing. The vast majority of usages are for single-row operations.

Also bring back a variety of different optimizations that (all in combination and with the changes above) gets an implementation of CoreCLR's `IMDInternalImport` for read-only usage to be almost 2% faster in my local benchmark.

I'm not bringing that interface into DNMD with this PR as that implementation also requires changes to `IMDInternalImport` to support DNMD's implementation (and as a result is not compatible with stock CoreCLR). Additionally, `IMDInternalImport`'s implementation is (as the name implies) internal to CoreCLR and not intended for public consumption.

When bringing DNMD into CoreCLR, I'll include the implementation with the requisite changes.

### CODE REVIEW

File: src/dnmd/access.c [hunk 1]
```diff
@@ -12,26 +11,95 @@ bool create_access_context(mdcursor_t* cursor, col_index_t col_idx, uint32_t row
...
         acxt->writable_data = NULL;
     }
-    
+    return true;
+}
+
+bool read_column_data(access_cxt_t* acxt, uint32_t* data)
+{
+    assert(acxt != NULL && acxt->data != NULL && data != NULL);
+    *data = 0;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/access.c:44]: No need for this. It should be set in all code paths.

File: src/dnmd/bytes.c [hunk 2]
```diff
@@ -159,6 +162,75 @@ bool read_i64(uint8_t const** data, size_t* data_len, int64_t* o)
...
 
+#else
+bool read_u16(uint8_t const** data, size_t* data_len, uint16_t* o)
+{
+    if (*data_len < 2)
+        return false;
+
+    memcpy(o, *data, 2);
+    *data += 2;
+    *data_len -= 2;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/bytes.c:173]:
```suggestion
    if (*data_len < sizeof(*o))
        return false;

    memcpy(o, *data, sizeof(*o));
    *data += sizeof(*o);
    *data_len -= sizeof(*o);
```

Using the above, should make all of the basically copy/paste.

File: src/dnmd/search.template.h [hunk 3]
```diff
@@ -10,6 +10,28 @@
 #error Must define SEARCH_FUNC_NAME(name) macro
 #endif // SEARCH_FUNC_NAME
 
+
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/search.template.h:13]:
```suggestion
```

File: src/dnmd/search.template.h [hunk 4]
```diff
@@ -21,7 +43,7 @@ static void const* SEARCH_FUNC_NAME(md_bsearch)(
     void* cxt)
 {
     assert(key != NULL && base != NULL);
-    while (count > 0)
+    while (count > 16)
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/search.template.h:46]: Why is this happening?

File: src/dnmd/query.c [hunk 5]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
+    if (!create_access_context(&c, col_idx, false, &acxt))
+        return false;
 
-    // If this isn't a #String column, then fail.
+    // If this isn't a heap index column, then fail.
     if (!(acxt.col_details & mdtc_hstring))
-        return -1;
+        return false;
+
+    mdstream_t const* heap = &
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:324]: What is this for?

File: src/dnmd/query.c [hunk 6]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
+    if (!create_access_context(&c, col_idx, false, &acxt))
+        return false;
 
-    // If this isn't a #US column, then fail.
+    // If this isn't a heap index column, then fail.
     if (!(acxt.col_details & mdtc_hus))
-        return -1;
+        return false;
+
+    mdstream_t const* heap = &
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:356]: What is this for?

File: src/dnmd/query.c [hunk 7]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
+    if (!create_access_context(&c, col_idx, false, &acxt))
+        return false;
 
-    // If this isn't a #Blob column, then fail.
+    // If this isn't a heap index column, then fail.
     if (!(acxt.col_details & mdtc_hblob))
-        return -1;
+        return false;
+
+    mdstream_t const* heap = &
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:389]: What is this for?

File: src/dnmd/query.c [hunk 8]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
-        return -1;
+    if (!create_access_context(&c, col_idx, false, &acxt))
+        return false;
 
-    // If this isn't a #GUID column, then fail.
+    // If this isn't a heap index column, then fail.
     if (!(acxt.col_details & mdtc_hguid))
+        return false;
+
+    mdstream_t const* heap = &
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:421]: ?

File: src/dnmd/query.c [hunk 9]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
+    acxt.table->cxt->strings_heap;
+    if (heap == NULL)
+        return false;
+
+#ifdef DEBUG_COLUMN_SORTING
+    validate_column_not_sorted(acxt.table, col_idx);
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:427]: This is new in all these functions, what is it for?

File: src/dnmd/query.c [hunk 10]
```diff
@@ -274,179 +266,225 @@ bool md_get_column_value_as_range(mdcursor_t c, col_index_t col_idx, mdcursor_t*
...
             return -1;
+
+        tk[read_in] = CreateTokenType(table_id) | table_row;
         read_in++;
     } while (out_length > 1 && next_row(&acxt));
 
     return read_in;
 }
 
+
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:487]:
```suggestion
```

File: src/dnmd/query.c [hunk 11]
```diff
@@ -506,14 +544,12 @@ static int32_t col_compare_2bytes(void const* key, void const* row, void* cxt)
...
-    assert(col_len == 2);
+    ASSERT_ASSUME(col_len == 2);
     bool success = read_u16(&col_data, &col_len, &rhs);
     assert(success && col_len == 0);
     (void)success;
 
-    return (lhs == rhs) ? 0
-        : (lhs < rhs) ? -1
-        : 1;
+    return lhs - rhs;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:552]:
```suggestion
    // 0: equal, < 0: (lhs < rhs), > 0: (lhr > rhs)
    return lhs - rhs;
```

File: src/dnmd/query.c [hunk 12]
```diff
@@ -526,14 +562,12 @@ static int32_t col_compare_4bytes(void const* key, void const* row, void* cxt)
...
-    assert(col_len == 4);
+    ASSERT_ASSUME(col_len == 4);
     bool success = read_u32(&col_data, &col_len, &rhs);
     assert(success && col_len == 0);
     (void)success;
 
-    return (lhs == rhs) ? 0
-        : (lhs < rhs) ? -1
-        : 1;
+    return lhs - rhs;
```
Comment (AaronRobinsonMSFT) [inline src/dnmd/query.c:570]:
```suggestion
    // 0: equal, < 0: (lhs < rhs), > 0: (lhr > rhs)
    return lhs - rhs;
```

**Review by AaronRobinsonMSFT — COMMENTED** [review]
I'm halfway done. There is some sloppiness here that is concerning. I see a great deal of code changes that seem to introduce new code paths that aren't clear why and it is hard to ensure I didn't miss anything.

File: src/dnmd/search.template.h (hunk 4 above)
Comment (jkoritzinsky) [reply src/dnmd/search.template.h:46]: This was from David's feedback (that also showed up in perf profiling) that switching to a linear search at a small number of elements ends up being better performance due to the branch predictor and the cache pre-fetching.

I'll add a comment here explaining that.

Comment (AaronRobinsonMSFT) [reply src/dnmd/search.template.h:49]: > that switching to a linear search at a small number of elements ends up being better performance due to the branch predictor and the cache pre-fetching.

This is slightly misleading. This logic only holds if `element_size` is also reasonably small or more precisely if the X * `element_size` can fit within the cache memory. Hardcoding 16 and not incorporating the `element_size` is precisely the sort of narrow optimization that can turn degenerate later on. I think we should factor that in or assert the element size is small enough not to matter.

**Pull request merged**